drivers/sony/displays/bravia.cr

```diff
@@ -0,0 +1,240 @@
+    cmd = Command.from_response?(parsed_data)
+    return task.try(&.abort("unrecognised command: #{parsed_data}")) if cmd.nil?
+    param = data[7..-1]
+    return task.try(&.abort("error")) if param.first? == MessageType::Error.value
+    case MessageType.from_value?(data[2])
+    when MessageType::Answer
+      update_status cmd, param
+      task.try &.success
+    when MessageType::Notify
+      update_status cmd, param
+    else
+      logger.debug { "Unhandled device response: #{data[2].chr rescue data[2]}" }
+      task.try &.abort("Unhandled device response")
+    end
+  end
+
+  COMMANDS = {
+    ir_code:           "IRCC",
+    power:             "POWR",
+    volume:            "VOLU",
+    audio_mute:        "AMUT",
+    mute:              "PMUT",
+    channel:           "CHNN",
+    tv_input:          "ISRC",
+    input:             "INPT",
+    toggle_mute:       "TPMU",
+    pip:               "PIPI",
+    toggle_pip:        "TPIP",
+    position_pip:      "TPPP",
+    broadcast_address: "BADR",
+    mac_address:       "MADR",
+  }
+
+  {% begin %}
+  enum Command
+    {% begin %}
+      {% for command in COMMANDS.keys %}
+        {{ command.camelcase.id }}
+      {% end %}
+    {% end %}
+
+    def function
+      {% begin %}
+      case self
+      {% for kv in COMMANDS.to_a %}
+        {% command, value = kv[0], kv[1] %}
+          in {{ command.camelcase }} then {{ value }}
+      {% end %}
+      end
+      {% end %}
+    end
+
+    def self.from_response?(message)
+      {% begin %}
+      case message
+        {% for kv in COMMANDS.to_a %}
+          {% command, value = kv[0], kv[1] %}
+          when {{ value }} then {{ command.camelcase.id }}
+        {% end %}
+      end
+      {% end %}
+    end
+  end
+  {% end %}
+
+  protected def convert_binary(data)
+    String.new(slice: data)
+  end
+
+  protected def request(command, parameter, **options)
+    cmd = command.function
+    parameter = parameter ? 1 : 0 if parameter.is_a?(Bool)
+    param = parameter.to_s.rjust(16, '0')
+    do_send(MessageType::Control, cmd, param, **options)
+  end
+
+  protected def query(state, **options)
+    cmd = state.function
+    do_send(MessageType::Enquiry, cmd, HASH, **options)
+  end
+
+  protected def do_send(type, command, parameter, **options)
+    cmd = "#{INDICATOR}#{type.control_character}#{command}#{parameter}\n"
+    send(cmd, **options)
+  end
+
+  protected def update_status(cmd : Command, param)
+    parsed_data = convert_binary(param)
+    case cmd
+    when .power?
+      self[:power] = parsed_data.to_i == 1
+    when .mute?
+      self[:mute] = parsed_data.to_i == 1
+    when .audio_mute?
+      self[:audio_mute] = parsed_data.to_i == 1
+    when .pip?
+      self[:pip] = parsed_data.to_i == 1
+    when .volume?
+      self[:volume] = parsed_data.to_i
+    when .mac_address?
+      self[:mac_address] = parsed_data.split('#')[0]
+    when .input?
+      self[:input] = Input.from_param(parsed_data[7..15])
+    end
```

Comment:
If you implemented a `#to_symbol` on the enum, you could do something like this
```suggestion
    state = case cmd
            when .power?, .mute?, .pip?
              parsed_data.to_i? == 1
            when .volume?
              parsed_data.to_i?
            when .mac_address?
              parsed_data.split('#').first?
            when .input?
              Input.from_param(parsed_data[7..15])
            end
    self[cmd.to_symbol] = state unless state.nil?
```

Reply:
I think things are good where they are, this just reduces the code a tiny bit

Reply:
wouldn't even need a to `to_symbol` it would be more like `#to_state_key : String` because you can index into a `NamedTuple` with a string